Documentation/Evolution/MatchingSemantics.md

```diff
@@ -0,0 +1,346 @@
+
+## Detailed Design
+
+The standard library should include a single `Regex` type, which by default matches at the `Character` level, using the same semantics as `String` equality comparisons. The `Regex` type provides API for changing those defaults, or you can use modifiers within a regular expression literal to change the matching semantics for portions of the expression. Swift regular expresssions have three modes that affect the matching semantics:
+
+- `Character` semantics (`X`)
+- Unicode scalar semantics (`u`)
+- Byte semantics (`b`)
+
+Enabling and disabling these modes have the behavior of pushing and popping from a stack. The compiler will warn when a mode has been ended (e.g. `(?-b)`) before it has been enabled.
+
+You can use any of these modes when applying a regular expression to a `String`, a `Substring`, or any collection of `UInt8`. When applied to any of these types, the collection will be interpreted at the specified semantic level, allowing a `String` to be interpreted as binary data, and allow grapheme-breaking and canonical equivalence algorithms to be applied to UTF-8 encoded bytes in a `Data` instance.
+
+Because the compiler determines the capture types of a regular epxression while parsing, we can preserve the ergonomics of captures that guarantee UTF-8 correctness. When a capture appears in a section of a regular expression literal that uses either `Character` or `UnicodeScalar` semantics, and that regular expression is applied to a `String` or `Substring`, the capture group is captured as a substring. However, if a capture includes any portion of a regular expression with byte semantics, or if a regular expression is applied to a non-string collection, then the capture group is captured as a collection that does not provide a UTF-8 correctness guarantee. See "Byte Semantics" for more detail.
+
+#### `Character` Semantics
+
+`Character`-based semantics mean that matching with a regular expression yields the same results as comparing for `String` equality. When matching with the default `Character`-level semantics, a "dot" or character class matches a single Swift `Character` and canonical equivalence is used when comparing:
+
+```swift
+let cafe = "Cafe\u{301}"      // "Café"
+cafe == "Café"                // true: String equality comparison
+cafe.matches(/Café/)          // true: canonical equivalence
+cafe.matches(/Cafe\u{301}/)   // true: character recognition in regex literals
+cafe.matches(/Caf./)          // true: dot matches a character
+cafe.matches(/Caf[åéîøü]/)    // true: canonical equivalence within custom classes
+cafe.matches(/Caf\w/)         // true: character class matches a character
+cafe.matches(/Caf\p{Letter}/) // true: Unicode property matches a character
+
+cafe.count == 4               // true
+cafe.matches(/.{4}/)          // true: dot matches a character
+
+// false: the fourth `.` matches the whole `"e\u{301}"` character
+cafe.matches(/....\u{301}/)
+```
```

Comment:
I think I need to break these longer examples into explanations of what's going on in more/less each line. For example, why `/Cafe\u{301}/` matches both `"Café"` and `"Cafe\u{301}"` when in `Character` semantic mode.